### ISSUE TITLE
Rework of the egg feature

### ISSUE THREAD

**Issue by NoxInmortus** [pterodactyl/panel#1906]
Hello,

First of all, allow me to repeat myself, but I like Pterodactyl a lot, it's almost all I ever wanted to manage game servers and give friendly admin access to my users.

For the context, allow me to add that I'm a Linux System Administrator, I work a lot on Docker, including a side-project with friends which have some common features with Pterodactyl (manage Docker containers through a web interface, but for a whole different usage).

I recently begun to adapt some Docker images to be Pterodactyl compliant, my Trackmania server for example, was already working very smooth on docker, but I wanted it to be managed through Pterodactyl. And the process... was and is PAINFUL (Actual work : https://github.com/NoxInmortus/docker-trackmania). Allow me to elaborate.

First of all this documentation is really spare on details : https://pterodactyl.io/community/config/eggs/creating_a_custom_image.html#creating-the-dockerfile , it does say that we have to be in `/home/container` with `container` user but not the why.

So after some time trying to figure out how the egg system works (it is honestly not very intuitive either) and some difficult labour, many tests and another issue (#1894), I was stuck. I had followed the documentation and Docker bests practices to obtain a fine image. But I couldnt make it work with Pterodactyl. Wings was constantly mounting an empty directory on `/home/container`and I couldnt figure it out why. Tried to play with `/mnt/server` either in several ways (but I had to dig this path out!!), without any luck.

For the whole time, I had the impression to be an archeologist trying to understand a really uselessly intricated process without understand why it was built this way, and it was really frustrating.

I finally had the opportunity to speak with @parkervcp, and I thanks him for his time again, he finally explained me that Wings force-mount an empty directory in `/home/container`, so that was not an issue on my side, but he also added details that helped me understand how the whole stuff was working, and why is was so much complicated.

Let me describe what I finally understand of a server creation (please correct me if anything is wrong to help me understand it fully) :
- A build phase is initiated, pulling the image, and running an entrypoint script, himself running the content of the install script of the "Install Script" part of an Egg. The install script must install the server content into `/mnt/server`
- Then, a second container is started (the Docker Image of the Egg), and the content of the `/mnt/server` of the first container is force-mounted into `/home/container` of the current one. And then, the entrypoint/cmd runs, eventually using the ${STARTUP} environnement variable (I really had to dig this one out).

And, to take the words of parkervcp, this is done this way because the game server runs as the `container` user without root-privileges and so cant do any install process which require root privileges. This is also done to ensure "Security on run" and "ease on install".

So, now that the context is fully explained, let me get to my point... apologies for being this long...

The goal of a Docker image is several. I will only quote some : to be easy to work on and to maintain, to be easily deployed, and to be more secure thanks to its isolation from its host. Well this is NOT the case here for the first two points.

I dont want to offend anyone, but it feels it went the wrong way. You dont have to do all thoses things to get a secured and easy to install container, I described the situation to demonstrate how much it is NOT easy to play with it.

As an example, my docker-trackmania image, does everything needing root privileges in the Dockerfile. Once builded, the entrypoint only setup game folder,files and run the game. It is faster, cleaner, and secured. You pull and run the image and that's it. And that's how it should be.

I am not saying to rework your entire egg system. But it should be able to run any docker image out-of-the-box. Just not overwriting `/home/container` and it would be enough. Not overwritting when "Skip install script" is checked would be the easiest thing to do. Using VOLUMES in the Dockerfiles and using automatic mounted points in /var/lib/docker/volumes would be easier and also probably necessary, instead of something in `/srv/daemon-data/`

I could go on on every detail of the egg system (I really could), and I could do it if you ask and bring my experience and skills to help you out because I REALLY LIKE Pterodactyl. But as I see it, the most pressing thing to do is to be able to run standard docker images. Just the maintenance would be easier, everything goes on the Docker image, and it would not need to use an external system with install script and second container locally built. Everything would be easier for EVERYONE, specially the dev.

I hope I wont get bashed out, I tried to bring earable arguments and I hope to bring a constructive discussion around core features. 

Thanks for your time.

**Comment by parkervcp** [comment]
The game files get stored outside the running container so we can have one docker container to run several servers. Instead of one image to run each game. Generic containers make it easy to manage multiple games.

Look at the source image that runs all the source based servers. The game files are stored out side the image to save space on images. So steam games don't need huge multi-gigabyte docker images per game.

Sorry I didn't have the time to explain that earlier

**Comment by NoxInmortus** [comment]
That's easily understandable, but my point still stands, being able to not-overwrite the `/home/container` would allow to use Docker images with a lot more ease and flexibility.

**Comment by parkervcp** [comment]
If you configure it properly the files don't get overwritten. If you install into the `/mnt/server/` folder (a docker volume bind mount) it gets remounted as `/home/container`.

**Comment by DaneEveritt** [comment]
Trying to run any arbitrary container on Pterodactyl is never going to work, and that is by design. We do lots of different filesystem actions _outside_ of Docker. Thus, allowing any image to be run would be nightmare inducing as we would have to be tracking all of the different locations files could end up, just so we could display and parse through them correctly.

Secondly, using docker volumes would introduce additional complexity when it comes to managing those files outside the container (e.g. through SFTP), and makes it difficult for novice users to appropriately backup or browse through the filesystem for their servers.

**Comment by NoxInmortus** [comment]
Hello,

Some more thoughts : 
> The game files get stored outside the running container so we can have one docker container to run several servers. Instead of one image to run each game. Generic containers make it easy to manage multiple games.
So instead of having multiple images you are having multiple install scripts. And when you're updating one egg, everyone wanting the update needs to either upgrade Pterodactyl, manually edit the Install script, or re-import an egg file. Instead of just pulling a new docker image tag.

> Look at the source image that runs all the source based servers. The game files are stored out side the image to save space on images. So steam games don't need huge multi-gigabyte docker images per game.
I did not though about that yesterday but...that's the entrypoint is made for. If you need to download large files, it should be in an included entrypoint script, so it wont enlarge the image. The current process allows that, to have the "install" part be run only once, that's a good point I concede.

> If you configure it properly the files don't get overwritten. If you install into the `/mnt/server/` folder (a docker volume bind mount) it gets remounted as `/home/container`.
Well, it seems to be the case through the "Install script" of an egg, but even if you install in `/mnt/server` within the dockerfile, theses files wont get mounted in `/home/container`. If I could only make this part work I would be enough for me and I will live with it.

> Thus, allowing any image to be run would be nightmare inducing as we would have to be tracking all of the different locations files could end up, just so we could display and parse through them correctly.
I dont understand how that would be more difficult to use standard docker behavior. You mount a folder in `/srv/daemon-data` exactly as you would have mounted something in `/var/lib/docker/volumes`. Of course, having multiple volumes will break the file management system (through gui and sftp), but what if you have only one docker volume ? Wont the process would be the exact same one but you would just need to get the docker mount id, and wont need all the actual process managing `/srv/daemon-data/` ?

**Comment by parkervcp** [comment]
You are arguing about putting the install process into the images and how the way we are doing it is backwards. But we are just approaching it from different sides. It's not a reason to attack the project. It's just a different solution to the same problem.

The install process is straight forward. The server files are kept on the host filesystem and not in the container.

The install process -  
1. create folder on host system `/srv/daemon-date/<server_uuid>`
2. start install container mounting `/srv/daemon-date/<server_uuid>` as `/mnt/server`
3. run install script
4. stop install container
5. start server container mounting `/srv/daemon-date/<server_uuid>` as `/home/container/`

**Comment by adaliszk** [comment]
Hmm, as a fresh person here I can understand the frustration that the documentation page does not really explain why things the way are, it does not really explain the individual modules that well and even the UI has some weird expectations to know that some fields are mandatory even though they are not marked like that.

On the other hand, there are some things what I cannot understand:

To have a building and then a runtime container which fits all type of servers for the specific architecture IS easy to work and maintain, it IS easy to deploy. Sure there are situations when you want to have a more tailored solution for cases when a runtime application requires extra packages, but then you can extend the main image anyway.

To have per game based images is totally doable, but you need a foundation for the various server-side architectures so the community could provide the more tailored images. A framework never tries to focus on one particular use-case, it tries to address the pattern.

This kind of approach is basically how, for example, lutris and wine works, there are base "images" in wine and there are custom-tailored images submitted by the community, but every installation has its own files on top of the "image" which is not submitted back to the repositories since it is custom for every user.

It is possible to create and use docker images for installation and for runtime. The install script could be just running a bash script which is included in the image in the `$.scripts.installation.container` property, and the runtime image is the main `$.image` which should be built from https://quay.io/pterodactyl/core. As described by **parkerpvc** the installation happens via a particular location within the docker image but outside it is the same location.

It might be an interesting feature request there to parameterize these with the egg configuration but honestly, it is much better if everybody uses the same pattern.  As far as I understand while the documentation does not guide you that deep in this regard, the things you want are possible already.

There are some features which would be great to have in regards to the design of the eggs for sure:
- Images to contain the install scripts (could be automated from the current eggs)
- Nest/Egg repositories to pull preconfigured setups and update them easily
- Server action to pull and rebuild [upgrade] the server (exposed to server owners too, possibly expose the available tags -> further extension of the eggs schema)
- Pre and Post upgrade script for the upgrade action where the build image would able to move files between the two installations
- Store runtime image tag for each server (to avoid sudden upgrade when a server rebuilds a container)
- Allow configuration of UDP and TCP port separately
- Extend egg configuration to set exposed ports on which the server creation form would know how much ports needed to be assigned to the server
- Extend the variables to be able to select ports into port variables from the allocated ports

I'm not sure if a complete rework is necessary at the moment, so far I only see missed opportunities that can be addressed in future requests and extending the documentation. For my use-case, I will need to build custom images, so along the way, I´m planning to open a few PRs to extend the functionality and the documentation a little bit.

**Comment by parkervcp** [comment]
- Why even maintain all the images when the install scripts are already doing the job? this would make even more images to maintain.
- I would be ecstatic if we could have a repo be in the panel. We would need to decide on a layout or format to support.
- Reinstall is available to users in 1.0 which is essentially an upgrade as it runs the install over the files generally updating them. At least in my eggs they would.
- pre-post upgrade scripts - how is an upgrade script related to moving files?
- store image tag - our entire  system is built on `repo:tag` as the image we would need to change versions. The way it's done now updates are applied to all server out there like rust.
- ports in the egg itself - I asked for this already - https://github.com/pterodactyl/panel/issues/1204
- apply both udp/tcp for ease to make sure it was never one or the other. (people are forgetful)
- In context to volumes - I asked for this and it was merged - https://github.com/pterodactyl/panel/pull/1915

**Comment by DaneEveritt** [comment]
Another important consideration for the entire design of this is that while it may seem strange from a more technologically inclined perspective, you have to also consider the _many_ users who rely on this software that have no clue how Docker works, what an image is, or how to install a new game.

Our implementation, while by no means the most perfect all-encompassing solution, does make it fairly easy for any average user to get a new game server running by selecting from our many base images, and just running and just coming up with a bash script to get their game installed.

I'd love to spend more time fleshing out this system on the backend to be a little less... _raw_ and more interactive, but right now the system works pretty well (and has done so for years now), so I'm not very inclined to make major breaking changes to it, especially given all of the other changes going on right now. :)

**Comment by adaliszk** [comment]
@parkervcp 
> Why even maintain all the images when the install scripts are already doing the job? this would make even more images to maintain.

Because of a pre-built image for each egg is a serious performance boost when it comes to installing a new server. Not only it would able to install faster but the common parts which do not need to build every single time would not affect the server performance while there is a new server installation.

The current way has a serious load when a new server is getting deployed and it often takes minutes to get into a state which could be a simple copy-paste from a pre-installed template image. An image download and extract is almost a zero load, so you could have a setup where you have a build server for new game versions and production servers where your hosted services just need to redeploy when there is an upgrade.

As for maintenance, you already spending time maintaining scripts, from your perspective it would not have any difference since the egg files can be parsed and build into images, what should be done is the separation of environmental preparation and game installation, maybe adding some version strings to generate the proper tags.

There are also things which should be explored like Kubernetes, multi-server/instance servers and integrate more into the Docker API with the egg schema, one day it might be deprecated and the project would only use Dockerfile behind the scenes, but that is a discussion for another day.

What would be certainly a huge help is to be able to locally build servers and test them out before publishing it onto production environments.

> store image tag - our entire system is built on `repo:tag` as the image we would need to change versions. The way it's done now updates are applied to all server out there like rust.

Yes, that is exactly the problem when you have a server that works with only one tag per environment version because you could have cases when the server works on a particular version and as soon there is an update it breaks some of the servers.

Right now, the way I can see this to work is to never upgrade an egg but import them into new versioned names each time but even then the main images don't have version tags to refer back to a previous state, so there must be some manual maintenance work to retag the working versions and maintain your own egg repository.

What I imagine here is to parse and build each egg into an image which would able to download the game resources, potentially via an onbuild script and a server-side docker-build for cache so then each deployment of an egg could be done in a matter of seconds and there would be the possibility to maintain servers with particular versions while new breaking issues are solved.

This whole process can be automated and tested, so things like latest tag would not be affected while a new version has some new requirements. An example could be that a game suddenly requires something new with a new version, but older installations are fine with the original base image.

> pre-post upgrade scripts - how is an upgrade script related to moving files?

If you have those hooks then each egg can define how to move files between the installations.

You could automate things like for example, in a Minecraft you could backup your worlds, plugin configurations before reinstalling the whole server, and post-scripts would able to pull the previous settings back and then restore the world.

To be fair, I'm a huge fan of CI/CD so if there would be some polish for creating any pipeline with simple YAML format, like in Gitlab-CI, then any kind of use-cases could be covered with less effort. You can keep it simple with only two stages (installation, deployment), or you could have a more sophisticated pipeline for complex actions.

> apply both udp/tcp for ease to make sure it was never one or the other. (people are forgetful)

Okay, that's actually not a bad approach not gonna lie, although this fact is not communicated through the UI that much. I can imagine security situations where it is important to guard connections with the protocol because there are existing methods to hack into services using one or the other protocol while the service is not prepared to serve on them.

@DaneEveritt 
> Another important consideration for the entire design of this is that while it may seem strange from a more technologically inclined perspective, you have to also consider the _many_ users who rely on this software that have no clue how Docker works, what an image is, or how to install a new game.

That is entirely fine, the users who don't know docker just need a good UI to guide them. For example, what's missing right now is a selectbox to select the common base images or use a custom one. Or any way to test an installation script without setting up and deleting servers.

It is also while space-efficient, not very good UX to have multiple columns because people are not likely to read into the second column, they tend to skip it.

> I'd love to spend more time fleshing out this system on the backend to be a little less... raw and more interactive, but right now the system works pretty well (and has done so for years now), so I'm not very inclined to make major breaking changes to it, especially given all of the other changes going on right now. :)

Totally understandable, and frankly most of the things are already exposed to exploring game-based images, so no need for extensive planning, it can be progressed with LEAN. The knowledge of how things work and why is the main issue in this ticket, and that one is covered quite well.

The UX/UI would help the most for the main users, the more technologically inclined people won't use an interactive UI anyway if they can just do a docker-build and see if the server is running or not on their local machine.

**Comment by NoxInmortus** [comment]
Hello,

I had time to think about this conversation.

@adaliszk glad to see I'm not alone with the same thoughts and needs, a lot of stuff you say makes sense and I just hope it wont be lost in the issue backlog.

To @DaneEveritt thanks for your time, I'm glad it is not ignored, I know things that may be interresting to changes would require some time to be thought and wrote.  As my original goal was to bring ideas and to make people think about it, that was fulfilled.

First, @parkervcp rest assured that it was not my intention to attack the project, and if it did look like that, my apologies. I just tried to bring my own insight and an external point of view of the situation, and I tend to speak and write with too much passion. I know understand much about how the project works, and I have some Trackmania and World of Warcraft eggs to write now.

I'm leaving this post as it is, but i'm sure we will talk again. Cheers

**Comment by DaneEveritt** [comment]
I'm going to close this out for the following reasons:

1.) I honestly just don't have the time to even begin approaching a problem such as this and also be able to push out features that I would deem as being significantly more important for stability and user experience.
2.) The current egg feature is such a fundamentally critical and defined part of this software that a change such as this would be an enormous undertaking and most likely cause extreme pain both on upgrade and support fronts.
